File: Week5/Practical/primesieve_skel.f90
module subs
    integer (kind=4) :: n
    integer (kind=4), allocatable :: primes(:)
    contains
      logical (kind=1) function testprime(i)
        implicit none
        integer (kind=4), intent(in) :: i 
        integer (kind=4) :: j,nsqroot

        nsqroot = ceiling(sqrt(real(i))) + 1
        if (nsqroot .ge. i) nsqroot = i - 1

        j = 2
        do while (mod(i,j).NE.0 .and. j.le.nsqroot) 
          j = j + 1
        end do

        testprime = .FALSE.    
        if (j .GT. nsqroot) testprime=.TRUE.
        return
      end function testprime

      subroutine removenums(i)
        integer (kind=4), intent(in) :: i
        integer (kind=4) :: j

        j = i*2
        do while (j .LE. n)
          primes(j) = -1
          j = j + i
        end do 
 
      end subroutine removenums
end module subs       



program primesieve
    use omp_lib
    use subs
    implicit none
    integer (kind=4) :: i,nprimes,ierr,tid


    write(6,*) ' What is the highest number you want to test?'
    read(5,*) n
    if (n .LT. 2) then
      write(6,*) ' Number must be bigger than 1, stopping'
      stop
    endif


    allocate(primes(n),stat=ierr)
    if (ierr .NE. 0) then
       write(6,*) ' Cannot allocate array, stopping '
       stop
    endif

    forall (i=1:n) primes(i)=i
    primes(1) = -1

end program primesieve

File: Week5/Practical/primesieve_skel.c
#include <stdio.h>
#include <omp.h>
#include <stdlib.h>
#include <math.h>

int n, *primes;

void removenums(int i);
int testprime(int i);

int main(void) {
    int i,nprimes,ierr,tid;


    printf(" What is the highest number you want to test?\n");
    scanf("%d",&n);
    if (n < 2) {
      printf("Number must be bigger than 1, stopping\n");
      exit(1);
    }

    primes = (int *) malloc(n*sizeof(int));

    if (primes == NULL) {
       printf(" Cannot allocate array, stopping\n");
       exit(2);
    }

    for (i=0; i<n; i++) primes[i] = i+1;
    primes[0] = -1;



}


int testprime(int i) {
    int  j,nsqroot;

/* Add one because array index added not number */
    i = i+1;

/* sqrt is the largest that needs to be checked */
    nsqroot = ceil(sqrt((float) i)) + 1;
    if (nsqroot >= i) nsqroot--;

    j = 2;
    while (i%j!=0 && j<=nsqroot) j++;
  
/* If not divisible by any of the numbers it must be prime */
    if (j > nsqroot) {
       return 0;
    } else {
       return 1;
    }
}

void removenums(int i) {
    int j;

    j = (i+1)*2;
    while (j < n) {
      primes[j-1] = -1;
      j = j + i+1;
    }
}